sdk/go/common/resource/plugin/langruntime_plugin.go

```diff
@@ -55,32 +55,102 @@ type langhost struct {
 // plugin could not be found, or an error occurs while creating the child process, an error is returned.
 func NewLanguageRuntime(host Host, ctx *Context, runtime, workingDirectory string, info ProgramInfo,
 ) (LanguageRuntime, error) {
-	path, err := workspace.GetPluginPath(ctx.Diag,
-		apitype.LanguagePlugin, strings.ReplaceAll(runtime, tokens.QNameDelimiter, "_"), nil, host.GetProjectPlugins())
+	attachPort, err := GetPluginAttachPort(runtime, PulumiDebugLanguagesEnvVar)
 	if err != nil {
 		return nil, err
 	}
 
-	contract.Assertf(path != "", "unexpected empty path for language plugin %s", runtime)
+	var plug *plugin
+	var client pulumirpc.LanguageRuntimeClient
+	if attachPort != nil {
+		port := *attachPort
 
-	args, err := buildArgsForNewPlugin(host, info.RootDirectory(), info.Options())
-	if err != nil {
-		return nil, err
-	}
+		handshake := func(
+			ctx context.Context, bin string, prefix string, conn *grpc.ClientConn,
+		) (*pulumirpc.LanguageHandshakeResponse, error) {
+			req := &pulumirpc.LanguageHandshakeRequest{
+				EngineAddress:    host.ServerAddr(),
+				RootDirectory:    "",
+				ProgramDirectory: "",
+			}
+			return languageHandshake(ctx, bin, prefix, conn, req)
+		}
 
-	plug, _, err := newPlugin(ctx, workingDirectory, path, runtime,
-		apitype.LanguagePlugin, args, nil, /*env*/
-		testConnection, langRuntimePluginDialOptions(ctx, runtime))
-	if err != nil {
-		return nil, err
+		conn, _, err := dialPlugin(
+			port,
+			"pulumi-language-"+runtime,
+			runtime+" (Language Plugin)",
+			handshake,
+			langRuntimePluginDialOptions(ctx, runtime),
+		)
+		if err != nil {
```

Comment:
I think we can assert here that dialPlugin _did_ return a result. Just will catch any stray cases of trying to debug an old language plugin that isn't supporting Handshake.